File: algorithms/binary_search/binary_search.js
var binarySearch = function(array, target) {
  if (array.length === 0 || (array.length === 1 && array[0] !== target)) {
    return -1;
  }
  var middle = Math.floor(array.length / 2);
  if (array[middle] === target) {
    return middle;
  }
  if (target < array[middle]) {
    return binarySearch(array.slice(0, middle), target);
  }
  var result = binarySearch(array.slice(middle + 1), target);
  return (result === -1) ? -1 : result + 1 + middle;
};

console.log(binarySearch([1, 4, 7, 8, 11, 13, 19], 13)); // Should be 5.

File: algorithms/binary_search/binary_search.rb
def binary_search(array, target)
  if array.empty? || (array.length == 1 && array.first != target)
    return nil
  end
  midpoint = array.length / 2
  return midpoint if array[midpoint] == target

  if target < array[midpoint]
    binary_search(array.take(midpoint), target)
  end

  result = binary_search(array.drop(midpoint + 1), target)
  result.nil? ? nil : result + midpoint + 1
end

if __FILE__ == $PROGRAM_NAME
  puts binary_search([1, 4, 7, 8, 11, 13, 19], 13) # Should be 5.
end
